src/reference_model/poseidon_python/finite_field.py

```diff
@@ -0,0 +1,171 @@
+        return (x << self.bit_num) % self.N
+
+    def __init__(self, value, N=p):
+        self.N = N
+        self.bit_num = int(math.log2(N) + 1)
+        self.R = pow(2, self.bit_num)
+        x, y = getxy(self.R, self.N)
+        self.N1 = -y
+        self.value = self.toMont(value)
+
+    def MonPro(self, op1, op2):
+        """Montgomery product"""
+        t = op1 * op2 % self.R
+        m = t * self.N1 % self.R
+        u = (op1 * op2 + m * self.N) >> self.bit_num
+
+        # print("ref intermediate_res:")
+        # print(hex(op1 * op2 + m * self.N))
+        while u >= self.N:
+            u -= self.N
+        return u
+
+    def ModAdd(self, op1, op2):
+        tmp = op1 + op2
+        while tmp > self.N:
+            tmp -= self.N
+        return tmp
+
+    def ModExp(self, x, exp):
+        """compute pow(x,exp) mod N in Montgomery pattern"""
+        exp_bin = bin(exp)[2:]
+        tmp = (1 * self.R) % self.N
+        for i in exp_bin:
+            tmp = self.MonPro(tmp, tmp)
+            if i == "1":
+                tmp = self.MonPro(x, tmp)
+        return tmp
+
+    def expassign(self, exp):
+        """self = pow(self,exp)"""
+        self.value = self.ModExp(self.value, exp)
+
+    def addassign(self, op1):
+        """self = self + op1"""
+        assert self.N == op1.N
+        self.value = self.ModAdd(self.value, op1.value)
+
+    def mulassign(self, op1):
+        """self = self * op1"""
+        assert self.N == op1.N
+        self.value = self.MonPro(self.value, op1.value)
+
+    def fromMont(self):
+        """transform Montgomery pattern into regular pattern"""
+        return self.MonPro(self.value, 1)
+
+    def add(self, op1):
+        """return op1 + self"""
+        assert self.N == op1.N
+        tmp = PrimeField(0, self.N)
+        tmp.value = self.ModAdd(self.value, op1.value)
+        return tmp
+
+    def mul(self, op1):
+        """return self * op1"""
+        assert self.N == op1.N
+        tmp = PrimeField(0, self.N)
+        tmp.value = self.MonPro(self.value, op1.value)
+        return tmp
+
+
+# test
+def testmul():
+    p1 = PrimeField(100)
+    p2 = PrimeField(999)
+    p3 = PrimeField(1234567890)
+
+    p1.mulassign(p2)
+    assert p1.fromMont() == (100 * 999) % p1.N
+    p2.mulassign(p3)
+    assert p2.fromMont() == (999 * 1234567890) % p2.N
+
+    print("pass mul test!!")
+
+
+def testadd():
+    a = 9870765432
+    b = 4567432198765430
+    c = 3421987654320
+    p1 = PrimeField(a)
+    p2 = PrimeField(b)
+    p3 = PrimeField(c)
+
+    p1.addassign(p2)
+    p2.addassign(p3)
+    assert p1.fromMont() == (a + b) % p1.N
+    assert p2.fromMont() == (b + c) % p1.N
+
+    print("pass add test")
+
+
+def testexp():
+    a = 65439870654
+    p1 = PrimeField(a)
+    p1.expassign(8)
+    assert p1.fromMont() == pow(a, 8) % p1.N
+
+    if p1.fromMont() != pow(a, 8) % p1.N:
+        print("false")
+    print("pass exp test!!")
+
+
+v = 0x11EFCB7C89C70F0124F7A5207AE9358022A7C2BEE13D4C79B5DCA2AE1C44CD1C

```

Comment:
global variables and functions in Python are not recommend, consider use __main__()

Reply:
fixed